Add support for Btrfs clone ioctl

szekelyszilv opened the pull request:
Btrfs clone ioctl can be used for O(1) copying within a file system
utilizing copy-on-write.

Passing --enable-btrfs-clone to ./configure enables this feature.

The feature is dependent on the "btrfs/ioctl.h" header which is part of
btrfs-progs.

vfs_file_task_do_copy is modified to always try to call the ioctl,
and fall back to the regular copying method if it fails.

szekelyszilv committed:
Add support for Btrfs clone ioctl

Btrfs clone ioctl can be used for O(1) copying within a file system
utilizing copy-on-write.

Passing --enable-btrfs-clone to ./configure enables this feature.

The feature is dependent on the "btrfs/ioctl.h" header which is part of
btrfs-progs.

vfs_file_task_do_copy is modified to always try to call the ioctl,
and fall back to the regular copying method if it fails.

szekelyszilv commented:
I found this useful e.g. for quickly and cheaply creating backup copies where a more robust approach was not required.

IgnorantGuru commented:
[Reference in coreutils cp](http://git.savannah.gnu.org/gitweb/?p=coreutils.git;a=blob;f=src/copy.c;h=191ccaaab5c5c7dd098950fbeb9023d83a61b924;hb=HEAD#l318)

Looks reasonable.  Need to document any build dependencies this would add.  Also configure should probably default to using it if the dependencies are present, like it does with startup-notification, since this seems standardized in coreutils.  If the deps aren't available, or if you add --disable-btrfs-clone, then it can skip it.  If you require --enable-btrfs-clone explicitly, it will rarely get used.

szekelyszilv committed:
Make Btrfs clone used by default

Change definition from USE_ to HAVE_

szekelyszilv committed:
Add Btrfs clone dependencies to README

szekelyszilv committed:
Add clone_file in VFS file tasks

Code cleanup, reduce #ifdefs
Add clone_file function that tries to clone using available methods in
turn. Return success if one of the succeeds or failure if none of them
do or are available.

szekelyszilv commented:
Changed it to be used by default and added it to the output of the configure script.
I also added to the README, dependencies are just btrfs-tools or btrfs-progs.
Also added the kernel and filesystem requirements, but those should fail gracefully.

In VFS file task the clone_file function can be used to add other clone methods
and have less conditional code.

nioncode commented:
Any chance of having this merged soon?
